lib/block_view_validator.go

```diff
@@ -0,0 +1,2133 @@
+package lib
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"github.com/deso-protocol/core/bls"
+	"github.com/dgraph-io/badger/v3"
+	"github.com/golang/glog"
+	"github.com/holiman/uint256"
+	"github.com/pkg/errors"
+	"io"
+	"math"
+	"net/url"
+	"sort"
+)
+
+//
+// TYPES: ValidatorEntry
+//
+
+type ValidatorEntry struct {
+	ValidatorID   *BlockHash
+	ValidatorPKID *PKID
+	// Note: if someone is updating their ValidatorEntry, they need to include
+	// all domains. The Domains field is not appended to. It is overwritten.
+	Domains                    [][]byte
+	DisableDelegatedStake      bool
+	VotingPublicKey            *bls.PublicKey
+	VotingPublicKeySignature   *bls.Signature
+	VotingSignatureBlockHeight uint64
+	TotalStakeAmountNanos      *uint256.Int
+	RegisteredAtBlockHeight    uint64
+	LastActiveAtEpochNumber    uint64
+	JailedAtEpochNumber        uint64
+	ExtraData                  map[string][]byte
+	isDeleted                  bool
+}
+
+func (validatorEntry *ValidatorEntry) Status() ValidatorStatus {
+	// ValidatorEntry.Status() is a virtual/derived field that is not stored in
+	// the database, but instead constructed from other ValidatorEntry fields.
+	// No sense in storing duplicative data twice. This saves memory and ensures
+	// that e.g. the ValidatorEntry.JailedAtEpochNumber field and the
+	// ValidatorEntry.Status() return value will never get out of sync.
+	//
+	// Make sure that any fields referenced here are included in the ValidatorMapKey
+	// since the ValidatorEntry.Status() value is used as a field in a Badger index.
+	if validatorEntry.JailedAtEpochNumber > uint64(0) {
+		return ValidatorStatusJailed
+	}
+	return ValidatorStatusActive
+}
+
+type ValidatorStatus uint8
+
+const (
+	ValidatorStatusInvalid ValidatorStatus = 0
+	ValidatorStatusActive  ValidatorStatus = 1
+	ValidatorStatusJailed  ValidatorStatus = 2
+)
+
+type ValidatorMapKey struct {
+	// The MapKey has to contain all fields that are used in Badger keys.
+	// Otherwise, an update to the UtxoView will not be able to update or
+	// delete all relevant Badger rows.
+	ValidatorPKID           PKID
+	TotalStakeAmountNanos   uint256.Int
+	RegisteredAtBlockHeight uint64
```

Comment:
Note to self: It's a bit weird that RegisteredAt is a block height but JailedAt is an epoch. Feels like one should shift to the other.